### ISSUE TITLE
`instance Ord a => Stream [a]`

### ISSUE THREAD

**Issue by tdammers** [mrkkrp/megaparsec#255]
This would be terribly useful for situations where you have a separate lexer, so that your parser becomes something like `[Lexeme] -> ParseTree`; but also for situations where your input is not a stream of bytes or characters, but, say, a stream of strings (e.g. for parsing CLI options). Most of the `String` instance, I believe could be reused, the only thing char-specific about it is the handling of newlines.

**Comment by mrkkrp** [comment]
I'm not sure how `defaultAdvance1` is going to work with such an abstract instance. If all we know about `a` is that it is an instance of `Ord`, I think we can't implement correct position-advancing methods of `Stream`. You're welcome to try though, I'm open to reviewing a PR that adds such an instance.

As for the use cases you mention, I would favor `newtype` wrappers for that.

**Comment by tdammers** [comment]
Well, the idea would obviously be that the position-advancing would simply increment the source column, and keep the source line at 1, since there is no notion of "newlines" for arbitrary tokens. This conflicts with the String instance though, so I agree that a newtype wrapper would be in order. Might still be useful to have a generic implementation of such a wrapper in the library though, after all this doesn't seem like an outrageous use case to me.

**Comment by mrkkrp** [comment]
If we just increment source column by 1 per token, wouldn't it be confusing? Because even with the examples you mention (a list of lexemes and a list of strings), a token (lexeme/string) is certainly capable of being "wider" than 1 column.

I think the problem here is that the library can parse things that are not tied to the notions of line and column, but the `SourcePos` type is currently defined so that this "system of coordinates" is imposed on everything we accept as input. Not sure what to do about that.

**Comment by tdammers** [comment]
So far, my mental model is such that the newtype wrapper would imply the semantics of source positions, and there are two solutions that make sense to me:

1. Every lexeme is considered exactly one unit wide, the lexeme stream is considered 1-dimensional, and "column" are considered in "lexeme space" (that is, column 24 refers to the 24th lexeme). This is easy, we'd just increment the column count by 1 for each lexeme. This is most useful for things like argument parsers, where we don't have 2-dimensional input anyway, and numbering lexemes matches the problem domain's semantics ("5th argument" makes sense for CLI commands, but line numbers really don't).
2. Some function is available that tells us the "size" of a lexeme, in terms of how many line breaks it introduces and how wide it is. Alternatively, we have a function that tells us the source position of each lexeme. Either way, this would allow the `advance` function to update the source position appropriately. The use case here would be to carry the lexeme's original source positions over into the parse tree, basically hiding the lexer step from the end user.

**Comment by mrkkrp** [comment]
I'm not sure I understand 2 correctly, but we already can do something like this, see:

https://github.com/mrkkrp/megaparsec/blob/master/tests/Text/MegaparsecSpec.hs#L1714-L1741

`Span` has its source position (start/end positions) inside and it's used by position advancing methods to set source position during parsing.

**Comment by tdammers** [comment]
> Span has its source position (start/end positions) inside and it's used by position advancing methods to set source position during parsing.

Yes, that's pretty much how I envision 2., except that it can be generalized to arbitrary token types, something like:

    data LexStream lexeme =
      LexStream
        { spanStart :: SourcePos
        , spanEnd :: SourcePos
        , spanBody :: lexeme
        }
        deriving (Read, Show, Eq, Ord, Functor)

And then the implementation of `Stream` is going to be almost the same, except that I think we don't need the `NonEmpty` stuff.

**Comment by alexanderkjeldaas** [comment]
I'm porting some code form regex-applicative, and this is biting me.

In my use-case I have no interest in getting parse errors at all, similarly to how a regexp parsing typically doesn't care about exact position information.

I would want option 1, not having to introduce a concept of the width of `a` being anything other than 1.
